decisions/DR013-Project-structure.md

```diff
@@ -0,0 +1,441 @@
+│                       └── python
+├── tests
+│   ├── openassetio-core
+│   │   ├── hostApi
+│   │   ├── managerApi
+│   │   └── trait
+│   ├── openassetio-core-c
+│   │   └── private
+│   │       ├── hostApi
+│   │       └── managerApi
+│   ├── openassetio-python
+│   └── python
+│       ├── openassetio
+│       │   ├── _core
+│       │   ├── hostApi
+│       │   │   └── resources
+│       │   ├── managerApi
+│       │   ├── pluginSystem
+│       │   │   └── resources
+│       │   │       ├── pathA
+│       │   │       ├── pathB
+│       │   │       │   └── packagedPlugin
+│       │   │       ├── pathC
+│       │   │       └── symlinkPath
+│       │   │           └── packagedPlugin -> ../pathB/packagedPlugin
+│       │   ├── test
+│       │   │   └── manager
+│       │   │       └── resources
+│       │   │           └── plugins
+│       │   └── traits
+│       └── openassetio_test
+├── thirdparty
+│   └── tomlplusplus
+│       └── include
+├── pyproject.toml
+├── setup.cfg
+└── setup.py
+```
+
+#### Pros
+
+ - No work to do.
+ - If/when supported, it's trivial to instruct adopters to simply
+  `pip install .` from the repo root.
+
+#### Cons
+
+- Gives the impression that `pip install .` from the repo root will
+  install a working OpenAssetIO library.
+- (Slightly) inconsistent directory structure for C/C++ components adds
+  cognitive complexity.
+- Duplication of `if (OPENASSETIO_ENABLE_<COMPONENT>)` under `tests`
+
+Estimated cost: Small
+
+
+### Option 2
+
+* Place pure Python component under `src/openassetio-python/python`
+  directory, including `setup.py` et al.
+* Move tests under the component that they are testing.
+  - Trickier for pure Python tests that cover multiple components. So
+    simply place alongside the Python package, i.e. in
+    `src/openassetio-python/python/tests`
+* Place `_openassetio_test` Python C++ extension module sources
+  alongside the `_openassetio` module.
+* Conform directory structure for C++ code with `include`, `src` and
+  `tests` subdirectories.
+* Place docs subdirectories under the repo-root `doc` directory (with
+  the addition of a `doxygen` subdirectory).
+
+```
+.
+├── cmake
+├── doc
+│   ├── contributing
+│   ├── decisions
+│   └── doxygen
+│       └── src
+├── resources
+│   ├── build
+│   ├── codegen
+│   │   ├── python
+│   │   │   └── openassetio_codegen
+│   │   │       ├── generators
+│   │   │       └── templates
+│   │   │           └── python
+│   │   ├── tests
+│   │   │   ├── generators
+│   │   │   └── resources
+│   │   ├── pyproject.toml
+│   │   └── setup.py
+│   ├── examples
+│   │   └── manager
+│   │       ├── BasicAssetLibrary
+│   │       │   ├── plugin
+│   │       │   │   └── BasicAssetLibrary
+│   │       │   ├── tests
+│   │       │   │   └── resources
+│   │       │   └── pyproject.toml
+│   │       └── SampleAssetManager
+│   │           ├── python
+│   │           │   └── SampleAssetManager
+│   │           └── tests
+│   └── perftest
+├── src
```

Comment:
[fix] Might be good to use an alternate word here - as we use `src` for c++ several layers down.

Reply:
@feltech and I went over this, and think leaving the doxygen folder where it is makes the most sense.

There were three viable options we considered : 
1. Move `doxygen` folder into `src` dir, treat it as an individual component
2. Split `doxygen` folder/docs build into individual components, and have a seperate `doxygen` folder under each component.
3. Leave where it is

1 - This makes some sense through an architectural lens. Like the python with pybind, docs are just another target using doxygen as a generator, so could be seen as a seperate component. However, the main argument against this, and in favour of keeping the folder in place, is simply intuitiveness. One would expect docs related items to be in `docs`, not `src`

2 - From the perspective of attempting to keep all modules trivially extractable, a fully developed forumulation may find `doxygen` folders in each component, (ie, generate different docs for core, c-api, and python.) This is challenging to at the moment however, due to the neccesity of generating component level documentation, and jigging up the build system.

3 - This makes the most sense to us, mainly due to intuitiveness, although we agree a super ideal system might use option #2, eventually.

Reply:
A top-level `src` directory is common across nearly all the reviewed examples from the ASFW Github org. Intuitively, I feel like that's the location where readers will expect to find our source code. 

Also if we assume all the components, other than core, are extracted, then the core would remain as the only `src` directory.

One option could be to do like OpenVDB and put all components at the top-level, rather than under an `src` subdirectory.  But that pollutes the root of the repo and is inconsistent if we wish to add other groupings (e.g. a `utils`, alongside `src`, with optional convenience utilities).

Another option would be to rename descendent `src` directories.  But then again, devs would expect to find source code under `src`.

Any suggestions?

Reply:
Chiming in to agree with David, we discussed this and don't think the reuse of `src` in the directory tree is misleading. Resolving this conversation in the absence of Tom.